File: windows/crates/router-core/tests/launcher_tests.rs
//! `SessionLauncher`: escolhe a conta que vai servir a sessão, ativa-a e devolve
//! o plano de lançamento. O processo em si fica na CLI — aqui é só a decisão,
//! testável sem lançar nada.
//!
//! Portados de `LauncherTests.swift` (suíte "SessionLauncher", 6), mais os dois
//! ramos da decisão que o Swift não fixava.

mod common;
use common::*;

use std::collections::HashMap;
use std::sync::Arc;

use router_core::engine::credential_store::CredentialStore;
use router_core::engine::provider::ProviderAdapter;
use router_core::engine::session_launcher::{LaunchError, SessionLauncher};
use router_core::{Account, AccountGroup, ConfigDir, Id, RouterConfig};

struct Fixture {
    launcher: SessionLauncher,
    config: RouterConfig,
    group: AccountGroup,
    a: Account,
    b: Account,
}

fn fixture() -> Fixture {
    let store = Arc::new(FakeStore::new());
    let adapter = Arc::new(FakeAdapter::new());
    let a = account("a@exemplo.com", "C:/Users/exemplo/.claude-a");
    let b = account("b@exemplo.com", "C:/Users/exemplo/.claude-b");
    store
        .write(&cred("ca"), &adapter.credential_location(&a.home))
        .unwrap();
    store
        .write(&cred("cb"), &adapter.credential_location(&b.home))
        .unwrap();
    let mut group = AccountGroup::new("Trabalho", ConfigDir::standard("C:/Users/exemplo"));
    group.account_ids = vec![a.id, b.id];
    group.threshold_percent = 90.0;
    let config = RouterConfig {
        accounts: vec![a.clone(), b.clone()],
        groups: vec![group.clone()],
        ..Default::default()
    };
    let launcher = SessionLauncher::new(store, vec![adapter as Arc<dyn ProviderAdapter>]);
    Fixture {
        launcher,
        config,
        group,
        a,
        b,
    }
}

fn usage(pairs: &[(Id, f64)]) -> HashMap<Id, f64> {
    pairs.iter().copied().collect()
}

#[test]
fn finds_a_group_ignoring_case_and_surrounding_spaces() {
    let f = fixture();
    assert_eq!(
        SessionLauncher::group_named(" trabalho ", &f.config).map(|g| g.name.as_str()),
        Some("Trabalho")
    );
    assert!(SessionLauncher::group_named("pessoal", &f.config).is_none());
}

#[test]
fn an_empty_group_is_a_named_error() {
    let launcher = SessionLauncher::new(
        Arc::new(FakeStore::new()),
        vec![Arc::new(FakeAdapter::new()) as Arc<dyn ProviderAdapter>],
    );
    let group = AccountGroup::new("g", ConfigDir::standard("C:/Users/exemplo"));
    let config = RouterConfig {
        groups: vec![group.clone()],
        ..Default::default()
    };

    let err = launcher
        .prepare(&group, &config, &HashMap::new(), Vec::new())
        .unwrap_err();
    assert_eq!(err, LaunchError::EmptyGroup("g".into()));
}

#[test]
fn starts_with_the_first_account_on_a_cold_start() {
    let f = fixture();
    let plan = f
        .launcher
        .prepare(&f.group, &f.config, &HashMap::new(), Vec::new())
        .unwrap();

    assert_eq!(plan.account.id, f.a.id);
    // O grupo padrão não exporta CLAUDE_CONFIG_DIR.
    assert_eq!(plan.config_dir_env, None);
    assert_eq!(plan.executable, "claude");
}

#[test]
fn keeps_the_active_account_while_it_has_headroom() {
    let f = fixture();
    f.launcher
        .prepare(&f.group, &f.config, &HashMap::new(), Vec::new())
        .unwrap(); // ativa `a`

    let plan = f
        .launcher
        .prepare(&f.group, &f.config, &usage(&[(f.a.id, 0.5)]), Vec::new())
        .unwrap();
    assert_eq!(plan.account.id, f.a.id);
}

#[test]
fn jumps_to_the_next_with_headroom_when_the_active_is_over() {
    let f = fixture();
    f.launcher
        .prepare(&f.group, &f.config, &HashMap::new(), Vec::new())
        .unwrap(); // ativa `a`

    let plan = f
        .launcher
        .prepare(
            &f.group,
            &f.config,
            &usage(&[(f.a.id, 0.95), (f.b.id, 0.1)]),
            vec!["--resume".into()],
        )
        .unwrap();
    assert_eq!(plan.account.id, f.b.id);
    assert_eq!(plan.arguments, vec!["--resume".to_string()]);
}

#[test]
fn a_dedicated_group_exports_claude_config_dir() {
    let store = Arc::new(FakeStore::new());
    let adapter = Arc::new(FakeAdapter::new());
    let a = account("a@exemplo.com", "C:/Users/exemplo/.claude-a");
    store
        .write(&cred("ca"), &adapter.credential_location(&a.home))
        .unwrap();
    let mut group = AccountGroup::new(
        "pessoal",
        ConfigDir::dedicated("C:/Users/exemplo/grupos/pessoal"),
    );
    group.account_ids = vec![a.id];
    let config = RouterConfig {
        accounts: vec![a],
        groups: vec![group.clone()],
        ..Default::default()
    };
    let launcher = SessionLauncher::new(store, vec![adapter as Arc<dyn ProviderAdapter>]);

    let plan = launcher
        .prepare(&group, &config, &HashMap::new(), Vec::new())
        .unwrap();
    assert_eq!(
        plan.config_dir_env.as_deref(),
        Some("C:/Users/exemplo/grupos/pessoal")
    );
}

// --- Os dois ramos que o Swift não fixava ---

/// Ninguém com folga: melhor subir na conta cheia do que recusar o lançamento
/// (o usuário pode trocar de modelo, ou só ler).
#[test]
fn stays_on_the_active_when_nobody_has_headroom() {
    let f = fixture();
    f.launcher
        .prepare(&f.group, &f.config, &HashMap::new(), Vec::new())
        .unwrap(); // ativa `a`

    let plan = f
        .launcher
        .prepare(
            &f.group,
            &f.config,
            &usage(&[(f.a.id, 0.95), (f.b.id, 0.97)]),
            Vec::new(),
        )
        .unwrap();
    assert_eq!(plan.account.id, f.a.id);
}

#[test]
fn an_account_that_cannot_be_activated_is_no_usable_account() {
    let launcher = SessionLauncher::new(
        Arc::new(FakeStore::new()),
        vec![Arc::new(FakeAdapter::new()) as Arc<dyn ProviderAdapter>],
    );
    let orphan = account("sem-credencial@exemplo.com", "C:/Users/exemplo/.claude-z");
    let mut group = AccountGroup::new("g", ConfigDir::standard("C:/Users/exemplo"));
    group.account_ids = vec![orphan.id];
    let config = RouterConfig {
        accounts: vec![orphan],
        groups: vec![group.clone()],
        ..Default::default()
    };

    let err = launcher
        .prepare(&group, &config, &HashMap::new(), Vec::new())
        .unwrap_err();
    assert_eq!(err, LaunchError::NoUsableAccount("g".into()));
}

File: macos/Tests/CCUsageCoreTests/LauncherTests.swift
import Foundation
import Testing
@testable import CCUsageCore

@Suite("SessionLauncher")
struct SessionLauncherTests {
    private func fixture() -> (SessionLauncher, RouterConfig, AccountGroup, Account, Account) {
        let kc = FakeKeychain()
        let adapter = FakeAdapter()
        let a = Account(provider: .anthropic,
                        identity: AccountIdentity(email: "a@k.com", organizationName: "K",
                                                  rateLimitTier: nil, raw: ["emailAddress": .string("a@k.com")]),
                        home: .dedicated("/tmp/.claude-a"))
        let b = Account(provider: .anthropic,
                        identity: AccountIdentity(email: "b@k.com", organizationName: "K",
                                                  rateLimitTier: nil, raw: ["emailAddress": .string("b@k.com")]),
                        home: .dedicated("/tmp/.claude-b"))
        try? kc.write("ca", service: adapter.keychainService(forConfigDir: a.home))
        try? kc.write("cb", service: adapter.keychainService(forConfigDir: b.home))
        let group = AccountGroup(name: "Trabalho", accountIDs: [a.id, b.id],
                                 configDir: .standard(home: "/Users/exemplo"), thresholdPercent: 90)
        let config = RouterConfig(accounts: [a, b], groups: [group])
        let launcher = SessionLauncher(keychain: kc, adapters: [adapter])
        return (launcher, config, group, a, b)
    }

    @Test("acha grupo sem diferenciar caixa nem espaços")
    func findsGroup() {
        let (launcher, config, _, _, _) = fixture()
        #expect(launcher.group(named: " trabalho ", in: config)?.name == "Trabalho")
        #expect(launcher.group(named: "pessoal", in: config) == nil)
    }

    @Test("grupo vazio é erro nomeado")
    func emptyGroup() {
        let kc = FakeKeychain()
        let launcher = SessionLauncher(keychain: kc, adapters: [FakeAdapter()])
        let group = AccountGroup(name: "g", accountIDs: [],
                                 configDir: .standard(home: "/Users/exemplo"))
        let config = RouterConfig(accounts: [], groups: [group])
        #expect(throws: SessionLauncher.LaunchError.emptyGroup("g")) {
            _ = try launcher.prepare(group: group, config: config, usage: [:], arguments: [])
        }
    }

    @Test("começa pela primeira conta quando nenhuma ativou ainda")
    func firstOnColdStart() throws {
        let (launcher, config, group, a, _) = fixture()
        let plan = try launcher.prepare(group: group, config: config, usage: [:], arguments: [])
        #expect(plan.account.id == a.id)
        // grupo padrão não exporta CLAUDE_CONFIG_DIR
        #expect(plan.configDirEnv == nil)
        #expect(plan.executable == "claude")
    }

    @Test("mantém a conta ativa enquanto ela tem folga")
    func keepsActiveWithHeadroom() throws {
        let (launcher, config, group, a, _) = fixture()
        _ = try launcher.prepare(group: group, config: config, usage: [:], arguments: [])  // ativa 'a'
        let plan = try launcher.prepare(group: group, config: config,
                                        usage: [a.id: 0.5], arguments: [])
        #expect(plan.account.id == a.id)
    }

    @Test("pula para a próxima com folga quando a ativa estourou")
    func rotatesOnLaunch() throws {
        let (launcher, config, group, a, b) = fixture()
        _ = try launcher.prepare(group: group, config: config, usage: [:], arguments: [])  // ativa 'a'
        let plan = try launcher.prepare(group: group, config: config,
                                        usage: [a.id: 0.95, b.id: 0.1], arguments: ["--resume"])
        #expect(plan.account.id == b.id)
        #expect(plan.arguments == ["--resume"])
    }

    @Test("grupo dedicado exporta CLAUDE_CONFIG_DIR")
    func dedicatedExportsEnv() throws {
        let kc = FakeKeychain(); let adapter = FakeAdapter()
        let a = Account(provider: .anthropic,
                        identity: AccountIdentity(email: "a@k.com", organizationName: nil,
                                                  rateLimitTier: nil, raw: [:]),
                        home: .dedicated("/tmp/.claude-a"))
        try? kc.write("ca", service: adapter.keychainService(forConfigDir: a.home))
        let group = AccountGroup(name: "pessoal", accountIDs: [a.id],
                                 configDir: .dedicated("/tmp/groups/pessoal"))
        let config = RouterConfig(accounts: [a], groups: [group])
        let launcher = SessionLauncher(keychain: kc, adapters: [adapter])
        let plan = try launcher.prepare(group: group, config: config, usage: [:], arguments: [])
        #expect(plan.configDirEnv == "/tmp/groups/pessoal")
    }
}

@Suite("ShellIntegration")
struct ShellIntegrationTests {
    @Test("statusLine aponta para o sensor")
    func statusLineShape() {
        let cmd = ShellIntegration.statusLineCommand(routerPath: "/Applications/X.app/router")
        #expect((cmd["command"] as? String)?.contains("statusline") == true)
        #expect((cmd["command"] as? String)?.contains("/Applications/X.app/router") == true)
    }

    @Test("instalar a statusLine preserva as outras chaves")
    func preservesSettings() throws {
        var written: Data?
        let existing = try JSONSerialization.data(withJSONObject: [
            "model": "claude-fable-5[1m]", "theme": "dark",
        ])
        try ShellIntegration.installStatusLine(
            routerPath: "/x/router",
            into: .dedicated("/tmp/g"),
            readFile: { _ in existing },
            writeFile: { data, _ in written = data })

        let root = try JSONSerialization.jsonObject(with: #require(written)) as! [String: Any]
        #expect(root["model"] as? String == "claude-fable-5[1m]")
        #expect(root["theme"] as? String == "dark")
        #expect((root["statusLine"] as? [String: Any])?["command"] != nil)
    }

    /// O caso que motivou a checagem: o `.app` foi renomeado de
    /// `ClaudeTokenCounter` para `FalcaoTokenRouter` em 18/09/2026 e as três
    /// peças da integração continuaram apontando para o caminho morto.
    @Test("statusLine de outro binário é obsoleta; a do binário atual não é")
    func statusLineStaleness() throws {
        let velho = try JSONSerialization.data(withJSONObject: [
            "statusLine": ["type": "command", "padding": 0,
                           "command": "'/Applications/ClaudeTokenCounter.app/Contents/MacOS/router' statusline"],
        ])
        #expect(ShellIntegration.statusLineIsStale(
            routerPath: "/Applications/FalcaoTokenRouter.app/Contents/MacOS/router",
            in: .dedicated("/tmp/g"), readFile: { _ in velho }))

        let novo = try JSONSerialization.data(withJSONObject: [
            "statusLine": ["type": "command", "padding": 0,
                           "command": "'/Applications/FalcaoTokenRouter.app/Contents/MacOS/router' statusline"],
        ])
        #expect(!ShellIntegration.statusLineIsStale(
            routerPath: "/Applications/FalcaoTokenRouter.app/Contents/MacOS/router",
            in: .dedicated("/tmp/g"), readFile: { _ in novo }))
    }

    /// Perfil sem `settings.json` nenhum conta como obsoleto: o sensor não está
    /// plantado ali, que é o mesmo resultado prático de apontar para o lugar
    /// errado — a conta fica sem medição e o rodízio decide às cegas.
    @Test("perfil sem statusLine conta como obsoleto")
    func missingStatusLineIsStale() {
        #expect(ShellIntegration.statusLineIsStale(
            routerPath: "/x/router", in: .dedicated("/tmp/g"), readFile: { _ in nil }))
    }

    @Test("a função de shell embute o caminho do binário")
    func shellFunctionHasPath() {
        let fn = ShellIntegration.shellFunction(routerPath: "/Applications/Falcão Router.app/router")
        #expect(fn.contains("claude()"))
        #expect(fn.contains("is-group"))
        #expect(fn.contains("launch"))
        #expect(fn.contains("Falcão Router.app/router"))
    }
}

@Suite("ShellIntegration profile")
struct ShellProfileTests {
    private func tempProfile(_ contents: String = "") -> URL {
        let url = URL(fileURLWithPath: NSTemporaryDirectory())
            .appending(path: "zshrc-\(UUID().uuidString)")
        try? contents.write(to: url, atomically: true, encoding: .utf8)
        return url
    }

    @Test("adiciona a linha uma vez e não duplica")
    func idempotent() throws {
        let profile = tempProfile("# meu zshrc\nexport PATH=/usr/bin\n")
        let script = "/Applications/X.app/Contents/MacOS/router"
        let line = "source \"\(script)/shell.sh\""

        let first = try ShellIntegration.ensureInProfile(
            sourceLine: line, scriptPath: "\(script)/shell.sh", profileURL: profile)
        #expect(first == .added)

        let second = try ShellIntegration.ensureInProfile(
            sourceLine: line, scriptPath: "\(script)/shell.sh", profileURL: profile)
        #expect(second == .alreadyPresent)

        let content = try String(contentsOf: profile, encoding: .utf8)
        // O conteúdo original ficou; a linha entrou exatamente uma vez.
        #expect(content.contains("# meu zshrc"))
        #expect(content.components(separatedBy: "shell.sh\"").count - 1 == 1)
    }

    @Test("cria o arquivo se ele não existe")
    func createsIfMissing() throws {
        let profile = URL(fileURLWithPath: NSTemporaryDirectory())
            .appending(path: "zshrc-missing-\(UUID().uuidString)")
        let out = try ShellIntegration.ensureInProfile(
            sourceLine: "source \"/x/shell.sh\"", scriptPath: "/x/shell.sh", profileURL: profile)
        #expect(out == .added)
        #expect(FileManager.default.fileExists(atPath: profile.path))
        try? FileManager.default.removeItem(at: profile)
    }
}

@Suite("ProfileSharing")
struct ProfileSharingTests {
    private func makeHome() -> (home: ConfigDir, base: URL) {
        let base = URL(fileURLWithPath: NSTemporaryDirectory())
            .appending(path: "share-\(UUID().uuidString)")
        let homeDir = base.appending(path: ".claude")
        let fm = FileManager.default
        try? fm.createDirectory(at: homeDir.appending(path: "projects"), withIntermediateDirectories: true)
        try? fm.createDirectory(at: homeDir.appending(path: "skills"), withIntermediateDirectories: true)
        try? "x".write(to: homeDir.appending(path: "history.jsonl"), atomically: true, encoding: .utf8)
        try? "y".write(to: homeDir.appending(path: "CLAUDE.md"), atomically: true, encoding: .utf8)
        return (ConfigDir.dedicated(homeDir.path), base)
    }

    @Test("linka projects e history quando o histórico é compartilhado")
    func linksHistory() throws {
        let (home, base) = makeHome()
        let group = ConfigDir.dedicated(base.appending(path: "grupo").path)
        ProfileSharing.link(into: group, home: home, shareHistory: true)

        let fm = FileManager.default
        for name in ["projects", "history.jsonl", "skills", "CLAUDE.md"] {
            let link = group.url.appending(path: name)
            #expect(fm.fileExists(atPath: link.path), "faltou \(name)")
            let isLink = (try? link.resourceValues(forKeys: [.isSymbolicLinkKey]))?.isSymbolicLink
            #expect(isLink == true, "\(name) devia ser symlink")
        }
        try? fm.removeItem(at: base)
    }

    @Test("sem histórico compartilhado, projects não é linkado")
    func skipsHistoryWhenOff() throws {
        let (home, base) = makeHome()
        let group = ConfigDir.dedicated(base.appending(path: "grupo").path)
        ProfileSharing.link(into: group, home: home, shareHistory: false)

        let fm = FileManager.default
        #expect(!fm.fileExists(atPath: group.url.appending(path: "projects").path))
        // mas skills (não é histórico) segue vindo
        #expect(fm.fileExists(atPath: group.url.appending(path: "skills").path))
        try? fm.removeItem(at: base)
    }

    @Test("não mexe no grupo padrão (seria linkar para si mesmo)")
    func skipsDefault() {
        ProfileSharing.link(into: .standard(home: "/tmp/whatever-xyz"), shareHistory: true)
        // Não cria nada em ~/.claude real; só não deve crashar.
        #expect(Bool(true))
    }
}

@Suite("ensureInProfile — integridade do ~/.zshrc")
struct ProfileIntegrityTests {
    /// Um `~/.zshrc` que não é UTF-8 válido. Acontece de verdade: um alias com
    /// acento salvo em latin-1, um arquivo tocado por um editor antigo.
    @Test("profile ilegível não pode ser sobrescrito")
    func doesNotDestroyUnreadableProfile() throws {
        let url = URL(fileURLWithPath: NSTemporaryDirectory())
            .appending(path: "zshrc-bin-\(UUID().uuidString)")
        // `alias café='echo oi'` com o "é" em latin-1 (0xE9) — byte inválido em UTF-8.
        var bytes = Data("alias caf".utf8)
        bytes.append(0xE9)
        bytes.append(contentsOf: Data("='echo oi'\nexport PATH=/meu/bin:$PATH\n".utf8))
        try bytes.write(to: url)
        let antes = try Data(contentsOf: url)

        _ = try? ShellIntegration.ensureInProfile(
            sourceLine: "source /x/shell.sh", scriptPath: "/x/shell.sh", profileURL: url)

        let depois = try Data(contentsOf: url)
        // O conteúdo original tem de continuar lá. Sobrescrever é perda de dados.
        #expect(depois.count >= antes.count,
                "o ~/.zshrc encolheu: o conteúdo do usuário foi destruído")
        #expect(depois.starts(with: antes),
                "o conteúdo original não foi preservado")
    }
}
